cmd/smashwords-downloader/main.go

```diff
@@ -98,36 +128,316 @@ func scrapeBookList(pageId int, dataDir string) {
 	bookCollector.OnHTML("div[id=pageContentFull]", func(e *colly.HTMLElement) {
 		title := e.ChildText("h1")
-		e.ForEach("a[title='Plain text; contains no formatting']", func(_ int, e *colly.HTMLElement) {
+		search := "a[title='Plain text; contains no formatting']"
+		if textFormat == "epub" {
+			search = "a[title='Supported by many apps and devices (e.g., Apple Books, Barnes and Noble Nook, Kobo, Google Play, etc.)']"
+		}
+
+		e.ForEach(search, func(_ int, e *colly.HTMLElement) {
 			book_link := e.Attr("href")
-			downloadBook(title, book_link, dataDir)
+			downloadBook(title, book_link, dataDir, textFormat)
 		})
 	})
 
-	westernRomanceURL := fmt.Sprintf("https://%s/books/category/%d/downloads/0/free/any/%d", smashWordsURL, westernRomanceID, pageId)
-	listCollector.Visit(westernRomanceURL)
+	smashwordsCategoryURL := fmt.Sprintf("https://%s/books/category/%d/downloads/0/free/any/%d", smashWordsURL, urlID, pageId)
+	listCollector.Visit(smashwordsCategoryURL)
 }
 
-func getDataDir() string {
-	dataDirPtr := flag.String("data_dir", "./data", "directory that the book files will download to")
+func main() {
+	//flags used: -url is the url to scrape,
+	// -data_dir is the directory to save the files to
+	dataDirPtr := flag.String("data_dir", "./data",
+		"directory that the book files will download to")
+
+	urlIDPtr := flag.Int("id", 1245,
+		"The cooresponding ID for the smashswords url you want to scrape"+
+			" (in https://www.smashwords.com/books/category/1245)")
+
+	itemsPerPagePtr := flag.Int("pageitems", 20,
+		"The number of items per page on the smashwords list page")
+
+	pagesPtr := flag.Int("pages", 7,
+		"The number of pages to scrape")
+
+	textFormatPtr := flag.String("format", "txt",
+		"The format of the book to download. Options are 'txt' or 'epub'")
+
+	overwriteSourcePtr := flag.Bool("overwriteSource", true,
+		"Save the original file after converting it to the desired format")
 	flag.Parse()
 
-	return *dataDirPtr
-}
+	totalBooks := *itemsPerPagePtr * *pagesPtr
 
-func main() {
-	dataDir := getDataDir()
+	//log the flag parameters out to console
+	log.Printf("Scraping %d pages of %d items, (total is %d) each from smashwords url %d.\n", *pagesPtr, *itemsPerPagePtr, totalBooks, *urlIDPtr)
+	log.Printf("Selected format is %s.\n", *textFormatPtr)
+	log.Printf("Saving files to %s folder.\n", *dataDirPtr)
 
+	// Create a wait group to wait for all the goroutines to finish
 	wg := new(sync.WaitGroup)
 
 	// Each list page only shows `bookListSize` books so scrape each one in parallel
-	for i := 0; i <= maxPageId; i = i + bookListSize {
+	for i := 0; i < (totalBooks); i = i + *itemsPerPagePtr {
 		wg.Add(1)
 		go func(pageId int) {
 			defer wg.Done()
-			scrapeBookList(pageId, dataDir)
+			scrapeBookList(pageId, *dataDirPtr, *urlIDPtr, *textFormatPtr)
 		}(i)
 	}
 
 	wg.Wait()
+
+	//convert epub to txt if needed
+	if *textFormatPtr == "epub" {
+
+		ConvertEpubGo(*dataDirPtr, *overwriteSourcePtr)
```

Comment:
This is already pretty fast, but if we want it to go even faster we can multi-thread the epub conversion.

Instead of the epub converter iterating through all the files in the data directory once everything's been downloaded, we could have the book scraper go routines write filenames that need to be converted through a channel. Then the epub converter goroutines read from that channel and do the conversion.